Data/ByteString/Lazy.hs

```diff
@@ -1435,43 +1435,56 @@ findIndexOrEnd k (S.BS x l) =
 -- * The program reads thousands of files, but due to lazy evaluation, the OS's file descriptor
 --   limit is reached before the handlers can be released.
 --
--- In order to avoid such unpleasant turn of events, results from lazy IO computation
--- must be fully evaluated.
+-- === Why?
 --
--- === __Dos and Don'ts of lazy IO__
+-- Consider the following program:
+--
+-- > import qualified Data.ByteString.Lazy as BL
+-- > main = do
+-- >   _ <- BL.readFile "foo.txt"
+-- >   BL.writeFile "foo.txt" mempty
+--
+-- Usual rules about 'IO' monad are that side effects happen
+-- sequentially and in full. That said, one would reasonably expect that
+-- reading a full file via 'readFile' executes all three actions
+-- (open a file handle, read content, close the file handle) before
+-- passing control to 'writeFile'. This expectation holds
+-- for strict "Data.ByteString" API. However, for lazy 'ByteString'
+-- the program above fails with "openBinaryFile: resource busy (file is locked)".
+--
+-- The thing is that "Data.ByteString.Lazy" is specifically designed
+-- to handle large or unbounded streams of data without requiring
+-- it to be resident in memory at once. This requirement cannot be satisfied
+-- if 'readFile' follows usual rules about 'IO': evaluating all side effects
+-- requires us to read file in full before closing its handle. This is why
+-- 'readFile' (and 'hGetContents' in general) is implemented
+-- via 'unsafeInterleaveIO', which bends the rules allowing 'IO' side effects
+-- to interleave on discretion of Haskell runtime. That's exactly what happens
+-- in the example above: 'readFile' opens a file handle, but since its result
+-- is not forced (and actually ignored),
+-- runtime assumes that side effects of reading contents
+-- and closing the file handle can be freely interleaved with further 'IO' effects,
+-- such as effects of 'writeFile'. So when 'writeFile' kicks in, @foo.txt@
+-- is still open for reading and cannot be opened for writing, causing
+-- the error quoted.
```

Comment:
```suggestion
-- Generally, in the 'IO' monad side effects happen 
-- sequentially and in full. Therefore, one might reasonably expect that
-- reading the whole file via 'readFile' executes all three actions
-- (open the file handle, read its content, close the file handle) before
-- control moves to the following 'writeFile' action. This expectation holds
-- for the strict "Data.ByteString" API. However, the above lazy 'ByteString' variant
-- of the program fails with "openBinaryFile: resource busy (file is locked)".
--
-- The reason for this is that "Data.ByteString.Lazy" is specifically designed
-- to handle large or unbounded streams of data incrementally, without requiring all the data
-- to be resident in memory at the same time. Incremental processing would not be possible
-- if 'readFile' were to follow the usual rules of 'IO': evaluating all side effects
-- would require reading file in full and closing its handle before returning from 'readFile'. This is why
-- 'readFile' (and 'hGetContents' in general) is implemented
-- via 'unsafeInterleaveIO', which allows 'IO' side effects to be delayed and
-- interleaved with subsequent processing of the return value.
-- That's exactly what happens
-- in the example above: 'readFile' opens a file handle, but since the content
-- is not fully consumed, the file handle remains open, allowing the content to
-- read __on demand__ (never in this case, since the return value is ignored).
-- So when 'writeFile' is executed next, @foo.txt@ is still open for reading and
-- the RTS takes care to avoid simultaneously opening it for writing, instead
-- returning the error shown above.
```

I hope the above is an improvement.  Just in case someone gets the idea that completely ignoring the return value is somehow important, we could show a somewhat more detailed example, in which the data is partly read, with similar results:
```haskell
import qualified Data.ByteString.Lazy.Char8 as BL
import Data.List (uncons)

main = do
    lines <- BL.lines <$> BL.readFile "/tmp/foo.txt"
    rest <- case uncons lines of
                Just (line1, rest) -> BL.putStrLn line1 >> return rest
                Nothing            -> return []
    -- Uncomment the line below to fully consume the input!
    -- mapM_ return rest
    BL.writeFile "/tmp/foo.txt" mempty
```
unless the input is fully consumed, resolving the immediate.

Of course we're then still left with the potential cost of holding all the data in memory if the results are consumed more than once!

Reply:
Thanks for proof reading, much appreciated. 

Not sure about additional examples: it feels to me that the length of this explanation already goes beyond haddock boundaries. I would love to link an article with more details (because actually there is not much specific to `bytestring` here), but cannot find anything decent.

Reply:
> Thanks for proof reading, much appreciated.
> 
> Not sure about additional examples: it feels to me that the length of this explanation already goes beyond haddock boundaries. I would love to link an article with more details (because actually there is not much specific to `bytestring` here), but cannot find anything decent.

I think the version you converged on is fine.  As for Haddock and lengths of text, my general beef with documentation systems of this sort is that they discourage writing expository prose of the sort that one sees in man page DESCRIPTION, NOTES, ... sections, focusing just on the function synopses.  However one can and should do both prose and brief per-function notes.  The idea is to do the longer write-ups separately at the head and/or foot of the module.  For an example, see what I did with [`Data.Traversable`](https://gitlab.haskell.org/ghc/ghc/-/blob/master/libraries/base/Data/Traversable.hs) or [pre-formatted](https://imrryr.org/~viktor/traversable/Data-Traversable.html)